### ISSUE TITLE
legion: slow startup at 96 nodes with shardrefine

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1539]
I've started scaling shardrefine on perlmutter and I see slow startup at 96 nodes. Things look ok before that, but at 96 nodes the start up time seems to take more than 10x. I ran it multiple times. 

http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/startup_legion_prof.html

Profiles are available here: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/

**Comment by lightsighter** [comment]
Did you run this in debug or release mode?

Did you get any warnings from the profiler about lots of long-latency messages?

Is there a way for me to run the 12-process version on sapling?

**Comment by lightsighter** [comment]
Can I get a profile for the 6 process one showing all the nodes?

**Comment by lightsighter** [comment]
FWIW, it seems like there is some really bad weirdness occurring when the Hand Off to MPI task runs and then it looks like the OS puts all the processes to sleep for a growing amount of time. Literally all execution stops on all the nodes, and the amount of time that takes scales with the number of nodes.

**Comment by syamajala** [comment]
This was RelWithDebInfo. 

I did not get any warnings from the profiler, but I think maybe unless you generate a full profile it doesnt give those warnings anymore? Will have to ask @elliottslaughter. 

I will see if the 6 process one runs on sapling on cpu.

There is a full profile from the 6 node run here: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/pwave_x_6_hept/legion_prof_full/

**Comment by elliottslaughter** [comment]
> I did not get any warnings from the profiler, but I think maybe unless you generate a full profile it doesnt give those warnings anymore?

Looking at the code, I believe the checks should be enabled all the time. I don't see any conditions for profiling a subset of nodes.

**Comment by lightsighter** [comment]
Ok, I see at least part of the problem (stupid tracing transitive reductions that take forever). I'd still like to have a 6-process test to run on sapling to play with to explore a couple of other things that look strange to me.

**Comment by syamajala** [comment]
The case is setup on sapling, but I'm seeing a new crash with the latest shardrefine. I guess there were some changes today and the one I ran on perlmutter was from Saturday. I will open a separate issue for that.

**Comment by lightsighter** [comment]
How do I run the case on sapling?

**Comment by lightsighter** [comment]
Please rebuild with the latest shardrefine as I at least fixed one of the issues.

**Comment by syamajala** [comment]
Did not see an improvement on perlmutter: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/startup_legion_prof.html

New profiles are here: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/

A full profile from 6 nodes is available as well: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/pwave_x_6_hept/legion_prof_full/

**Comment by lightsighter** [comment]
These profiles are different. The Realm DMA channels are just getting completely swamped. The utility processors no longer look like the bottleneck. The copies were probably the limiter before, but it's even more obvious now that I fixed the most egregious overheads on the utility processors. FWIW, the reason the utility processors look a bit slow at times is because Realm is so swamped with data movement.

**Comment by lightsighter** [comment]
I guess this is less of a problem on the 48 and 96 node runs, something else is going on there.

**Comment by lightsighter** [comment]
I would still like a way to run the 6-node version on sapling.

**Comment by lightsighter** [comment]
Please regenerate the 96-node profile with the latest `shardrefine` as it has more detailed information about the messages being sent in the profiler.

**Comment by syamajala** [comment]
A new profile is available here: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/pwave_x_96_hept/legion_prof/

**Comment by syamajala** [comment]
If you want to run the 6 rank case on sapling you can do so now.

```
salloc -N 1 -p cpu --exclusive
cd /scratch2/seshu/legion_s3d_subranks/Ammonia_Cases
./ammonia_job.sh
```

**Comment by lightsighter** [comment]
Can you try making your projection functions pure functional ones so that Legion can memoize the projection trees that it has to make? It looks like we're doing a ton of redundant communication because Legion can't memoize projection function results. There is an inherent all-gather operation that Legion has to do every time it sees a projection functor invoked from a new region/partition for the first time. The data structure we create out of this all-gather is scalable (it's size is independent of the size of the number of regions used in the projection), but the amount of data exchanged in the all-gather is not really scalable (it grows with the number of subregions used in the projection). If we can memoize the result then we don't have to do it every time like we do right now in this application. If we can't mark the projection functions as being memoizable then I think we're going to need to just fall back to a super pessimistic logical analysis that just assumes that anything that interferes on the roots of the projection interfere in their sub-region usage as well which is going to necessitate merge close op fences all over the place.

**Comment by elliottslaughter** [comment]
I'm not sure we documented this decision, but we need the interface you suggested in https://github.com/StanfordLegion/legion/issues/1410#issuecomment-1459932407. The reason why the Regent projection functors aren't functional via the current interface is described in https://github.com/StanfordLegion/legion/issues/1410#issuecomment-1452651565.

**Comment by lightsighter** [comment]
> I'm not sure we documented this decision, but we need the interface you suggested in https://github.com/StanfordLegion/legion/issues/1410#issuecomment-1459932407.

The current implementation in `shardrefine` takes into account memoizing the "[state data](https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h?ref_type=heads#L1107-1108)" passed in on the region requirement, so if that is all you are accessing then you can safely mark the projection functor as "pure functional" and allow Legion to memoize the result.

**Comment by lightsighter** [comment]
If that is the only problem, then I suppose we can add a new project method and give you a mechanism to say in the projection functor which interface gets called. What should the project method look like?

```
class ProjectionFunctor {
 ...
  virtual LogicalRegion project(LogicalRegion upper bound, const DomainPoint &point, const Domain &launch_domain, const void *args, size_t arglen);
  virtual LogicalRegion project(LogicalPartition upper_bound, const DomainPoint &point, const Domain &launch_domain, const void *args, size_t arglen);
```

Alternatively, we can not add a new interface, I give you a way to just promise that the result is functional on those parameters and trust that you're doing the right thing even though we're calling the older interface.

**Comment by elliottslaughter** [comment]
That interface is fine with me.

I seem to recall you also wanted to deprecate and remove the old non-functional interface. But that is not important to me.

**Comment by lightsighter** [comment]
Here is the new interface:

https://gitlab.com/StanfordLegion/legion/-/commit/9dbebae824e3eefa65ee6921e4969d017eec9f62

It will automatically use the new methods if you mark the projection functor as pure functional and you region requirement has projection arguments. Unfortunately there is no way to back-port this before `shardrefine`. The machinery is just not there in the runtime to support it.

**Comment by elliottslaughter** [comment]
Does this C API look ok?

https://gitlab.com/StanfordLegion/legion/-/commit/8c0c088affa6e661c881cad63df1fc1a55d4d24a

**Comment by elliottslaughter** [comment]
Actually, I keep needing to make fixes, so let's do this via MR: https://gitlab.com/StanfordLegion/legion/-/merge_requests/925

**Comment by elliottslaughter** [comment]
Ok, that MR now has Regent support too, though it's crashing at runtime. Either (a) I failed to correctly register the right kind of projection functor, or (b) the runtime is not calling the correct method.

**Comment by lightsighter** [comment]
Something about the projection functor arguments didn't make sense to me. They are supposed to be dynamic in the sense that each new task launch might annotate new arguments on each of its region requirements. Legion is memoizing over the binary value of those arguments. Therefore they shouldn't be member variables on the functor object, but instead something that gets passed through to the `project` function call.

I do see the CI failing, but I don't know how to interpret the assertion.

**Comment by elliottslaughter** [comment]
That's just how we wrap the C++ API to offer the C API. The members are the pointers to the C API function pointers that are the true projection functors. This is not new, it's how we've always done projection functors in C.

I haven't dug into debugging yet, but glancing through the code I don't see an obvious mismatch on my side. One way or another, it appears that the wrong projection functor method is getting called.

**Comment by elliottslaughter** [comment]
This is my issue. I'll follow up with a fix.

**Comment by elliottslaughter** [comment]
Test pass locally. Feel free to review the MR again. All my changes were inside of the compiler.

**Comment by lightsighter** [comment]
Looks good to me, merge when ready. Then we can get @syamajala to try again with the updated Regent code.

**Comment by elliottslaughter** [comment]
Merged to `shardrefine`. @syamajala please try it out and see what you get.

**Comment by syamajala** [comment]
Runs are in the queue. Some of them have gone through and it at least looks like it runs without crashing.

**Comment by syamajala** [comment]
Nothing past 12 nodes seemed to work correctly. 24 and 48 nodes got stuck after 10 timesteps, 96 nodes never finished the first timestep. I submitted some runs again.

**Comment by lightsighter** [comment]
Run with -ll:force_kthreads and dump back traces from all nodes when they
hang.

**Comment by syamajala** [comment]
Its on Perlmutter, which is slingshot-11. I'm running with Realm flow control on top of shardrefine.

**Comment by syamajala** [comment]
The runs I resubmitted worked through 48 nodes, but 96 doesnt start up after half an hour. Will have to see if I can get stack traces.

**Comment by syamajala** [comment]
I have stacktraces from a 96 node job here: http://sapling2.stanford.edu/~seshu/s3d_perlmutter_subrank/stacktraces/

I took them at 4 different points over the course of half an hour.

**Comment by lightsighter** [comment]
There's definitely something strange going on. At least between the last two sets of backtraces nothing changed so it is properly stuck and things are not just running slowly. I didn't check the early backtraces.

Every node has a non-trivial backtrace with at least one thread stuck in the logical dependence analysis waiting on messages from remote nodes. Different nodes have varying number of outstanding operations stuck in the prepipeline stage also waiting on messages from remote nodes. This looks to me like the network is getting stuck like before. Do you want to go through the trouble of confirming that messages are getting stuck?

**Comment by syamajala** [comment]
Its clear that this is a network issue. I ran without Realm flow control and just set the libfabric environment variables and it started up. Maybe thats fine for now though. Running rank/node uses significantly less memory so we can sort of afford to let libfabric go wild. 

Long term though, Slingshot 2.1 should be starting to get rolled out to machines soon, and from the initial tests we've heard from the gasnet guys and OLCF it seems to be working. I don't know when it will become available on Perlmutter though, but Frontier should be getting it any day.

**Comment by syamajala** [comment]
Slingshot 2.1 did fix these issues. See #1581.
